runtime-sdk/src/modules/rofl/mod.rs

```diff
@@ -0,0 +1,369 @@
+    NotSignedByRAK,
+
+    #[error("unknown enclave")]
+    #[sdk_error(code = 4)]
+    UnknownEnclave,
+
+    #[error("unknown node")]
+    #[sdk_error(code = 5)]
+    UnknownNode,
+
+    #[error("endorsement from given node not allowed")]
+    #[sdk_error(code = 6)]
+    NodeNotAllowed,
+
+    #[error("registration expired")]
+    #[sdk_error(code = 7)]
+    RegistrationExpired,
+
+    #[error("extra key update not allowed")]
+    #[sdk_error(code = 8)]
+    ExtraKeyUpdateNotAllowed,
+
+    #[error("core: {0}")]
+    #[sdk_error(transparent)]
+    Core(#[from] modules::core::Error),
+}
+
+/// Gas costs.
+#[derive(Clone, Debug, Default, cbor::Encode, cbor::Decode)]
+pub struct GasCosts {
+    pub tx_register: u64,
+    pub internal_is_authorized_origin: u64,
+}
+
+/// Parameters for the module.
+#[derive(Clone, Debug, Default, cbor::Encode, cbor::Decode)]
+pub struct Parameters {
+    pub gas_costs: GasCosts,
+}
+
+/// Errors emitted during rewards parameter validation.
+#[derive(thiserror::Error, Debug)]
+pub enum ParameterValidationError {}
+
+impl module::Parameters for Parameters {
+    type Error = ParameterValidationError;
+
+    fn validate_basic(&self) -> Result<(), Self::Error> {
+        Ok(())
+    }
+}
+
+/// Genesis state for the rewards module.
+#[derive(Clone, Debug, Default, cbor::Encode, cbor::Decode)]
+pub struct Genesis {
+    pub parameters: Parameters,
+}
+
+pub struct Module<Cfg: Config> {
+    _cfg: std::marker::PhantomData<Cfg>,
+}
+
+#[sdk_derive(Module)]
+impl<Cfg: Config> Module<Cfg> {
+    const NAME: &'static str = MODULE_NAME;
+    type Error = Error;
+    type Event = ();
+    type Parameters = Parameters;
+    type Genesis = Genesis;
+
+    #[migration(init)]
+    fn init(genesis: Genesis) {
+        genesis
+            .parameters
+            .validate_basic()
+            .expect("invalid genesis parameters");
+
+        // Set genesis parameters.
+        Self::set_params(genesis.parameters);
+    }
+
+    /// Register a new ROFL instance.
+    #[handler(call = "rofl.Register")]
+    fn tx_register<C: Context>(ctx: &C, body: types::Register) -> Result<(), Error> {
+        let params = Self::params();
+        <C::Runtime as Runtime>::Core::use_tx_gas(params.gas_costs.tx_register)?;
+
+        if body.expiration <= ctx.epoch() {
+            return Err(Error::RegistrationExpired);
+        }
+
+        let policy = Cfg::auth_policy();
+        let app_policy = policy.apps.get(&body.app).ok_or(Error::UnknownApp)?;
+
+        if body.expiration - ctx.epoch() > app_policy.max_expiration {
+            return Err(Error::InvalidArgument);
+        }
+
+        // Ensure that the transaction is signed by RAK.
+        let caller_pk = CurrentState::with_env(|env| env.tx_caller_public_key())
+            .ok_or(Error::NotSignedByRAK)?;
+        if caller_pk != body.ect.capability_tee.rak {
+            return Err(Error::NotSignedByRAK);
+        }
+
+        if CurrentState::with_env(|env| env.is_check_only()) {
```

Comment:
This is to avoid doing too much work during checks where fees have not yet been paid.

Reply:
> This is to avoid doing too much work during checks where fees have not yet been paid.

Since this is probably a general guideline, we don't need to add this as a comment. Resolving.